### ISSUE TITLE
clair.log grows rapidly with the same message on repeat - scan loop?

### ISSUE THREAD

**Issue by darkl0rd** [goharbor/harbor#7316]
After an upgrade from 1.6.0 to 1.7.5 I've noticed that my clair log has the following message on repeat:

Apr  7 20:57:58 172.18.0.1 clair[58049]: {"Event":"Handled HTTP request","Level":"info","Location":"router.go:57","Time":"2019-04-07 18:57:58.244374","elapsed time":889528680,"method":"GET","remote addr":"172.21.0.5:57596","request uri":"/v1/layers/c6faafb4bc1db7687604be8b4978c0fbf5e75f71bb20f87cadc1998e1344ab69?features\u0026vulnerabilities","status":"200"}

This results in massive logs; but I'm also under the impression that this is causing clair to keep restarting scans; I've already seen a couple of times that it reported max connections on the postgres database.

**Comment by darkl0rd** [comment]
see also #7028 which seems to describe the same issue without a resolution, yet closed his issue.

**Comment by reasonerjt** [comment]
@darkl0rd could you attach the complete log file of clair and postgres?

**Comment by darkl0rd** [comment]
I have just deleted all logs, restarted the instance. Issue reproduces, postgres log remains quiet for quite a bit (so have not attached that), the ones that grow quickly are core.log and clair.log which are attached.

[harbor-logs.zip](https://github.com/goharbor/harbor/files/3052785/harbor-logs.zip)

**Comment by reasonerjt** [comment]
Hi it seemed that in your env Clair failed to send a lot notifications for vuln update, and Once Harbor receives such notification, it will try to re-scan all images, this should finish relatively fast because Clair already scanned these images so no disk I/O really happens.

Are the log files keep growing after 2 days?

**Comment by kumy** [comment]
Coming from #7028, the logs loop never ended after many days. I end-up building a fresh instance of harbor and migrated my data there manually. (It was a lots of pain)

**Comment by reasonerjt** [comment]
@kumy you gist is gone, was it also due to the notifications?  based on the logs in original thread the notifications were removed after handled so I assume it will stop sending notification as some point.

**Comment by reasonerjt** [comment]
I think the way Harbor handles notification is a little problematic, I will think about how to refine it.

**Comment by kumy** [comment]
@reasonerjt sorry I don't have the logs anymore.

**Comment by darkl0rd** [comment]
@reasonerjt The problem persists, logs keep filling up with the same message continuously:

-rw-r--r-- 1 10000 10000  55M Apr 10 13:25 core.log
-rw-r--r-- 1 10000 10000 178M Apr 10 13:25 clair.log

**Comment by reasonerjt** [comment]
Could you zip the logs and see if thats possible to attach to this issue?  
if not could you upload it to an online storage?

**Comment by darkl0rd** [comment]
Sure, but it's more of what I posted earlier ...

**Comment by darkl0rd** [comment]
[clair.zip](https://github.com/goharbor/harbor/files/3063732/clair.zip)

**Comment by amonhuang** [comment]
I have same issue too in 1.7.5, these message still pump out when the service stack restart. 
Here is my clair.log:

> Apr 24 11:50:44 172.18.0.1 clair[1034]: {"Event":"found and locked a notification","Level":"info","Location":"notifier.go:143","Time":"2019-04-24 03:50:44.816669","notification name":"d3c3af8c-fb30-4ec6-85e2-0facc3ef21c1"}
Apr 24 11:50:44 172.18.0.1 clair[1034]: {"Event":"Handled HTTP request","Level":"info","Location":"router.go:57","Time":"2019-04-24 03:50:44.823813","elapsed time":6190432,"method":"GET","remote addr":"172.20.0.4:45856","request uri":"/v1/notifications/d3c3af8c-fb30-4ec6-85e2-0facc3ef21c1?limit=2","status":"200"}
Apr 24 11:50:44 172.18.0.1 clair[1034]: {"Event":"Handled HTTP request","Level":"info","Location":"router.go:57","Time":"2019-04-24 03:50:44.827397","elapsed time":1266474,"method":"DELETE","remote addr":"172.20.0.4:45856","request uri":"/v1/notifications/d3c3af8c-fb30-4ec6-85e2-0facc3ef21c1","status":"200"}
Apr 24 11:50:44 172.18.0.1 clair[1034]: {"Event":"successfully sent notification","Level":"info","Location":"notifier.go:184","Time":"2019-04-24 03:50:44.828080","notification name":"d3c3af8c-fb30-4ec6-85e2-0facc3ef21c1"}
...

**Comment by kumy** [comment]
I remember that during the log spam phases, the database connections increased haphazardly to finally reach the maximum database connection.

**Comment by darkl0rd** [comment]
After upgrading to v1.8.0, clair is not working at all anymore.

On the Vulnerability page I get an "Internal server error" and when attempting to edit the schedule it also throws an error. The logs show nothing sensible.

**Comment by greg-dobedobedo** [comment]
Hi @darkl0rd, I'm facing the same problem. 
Could you tell how you clear out the jobs in the database please ?

**Comment by reasonerjt** [comment]
@darkl0rd sorry for the late response, the way that Harbor handles the webhook from Clair is problematic when there are many many images.

We are considering to remove it.

To workaround this problem you can comment out this section from `config.yaml` of clair, ( `harbor/common/config/clair/config.yaml`):
```
  notifier:
    attempts: 3
    renotifyinterval: 2h
    http:
      endpoint: http://core:8080/service/notifications/clair
```

and restart Harbor.

**Comment by greg-dobedobedo** [comment]
@reasonerjt, Thank you for this workaround. Until now, no loop anymore.

**Comment by caparisi** [comment]
Same issue on 1.4. After upgrade to 1.8.1 this is issue persist, but the workaround fix it (thanks !). We are looking forward to version 1.9.0 !

**Comment by reasonerjt** [comment]
This should be fixed after #8466 is merged.  Closing.
